### PR TITLE
Templatize value-set analysis

### PR DESCRIPTION
**Pull request by smowton**
A second attempt at https://github.com/diffblue/cbmc/pull/1413 -- see there for detailed description.

Differences vs. that PR:

* Name `value_set_domaint` retained as-is, and the templated version of the same class given a fresh name.
* Kept as similar as possible a file layout to the pre-existing code, to shrink the diff and make merge conflicts easier to resolve.

### CODE REVIEW

**Comment by smowton** [comment]
In comments on #1413 @polgreen confirms this is no longer a breaking change for her derived project. Would be good to find others willing to test it out if possible.

**Comment by smowton** [comment]
@tautschnig any opinion on this? It should be the same as #1413, which was accepted but then reverted for breaking outside projects, but slightly restructured to minimise churn.

File: src/goto-symex/postcondition.cpp [hunk 1]
```diff
@@ -174,7 +174,7 @@ bool postconditiont::is_used(
     // aliasing may happen here
 
     value_setst::valuest expr_set;
-    value_set.get_value_set(expr.op0(), expr_set, ns);
+    value_set.read_value_set(expr.op0(), expr_set, ns);
```
Comment (tautschnig) [inline src/goto-symex/postcondition.cpp]: I'm only starting to review this, so please forgive me if I'm asking a dumb question: is this renaming essential?

File: src/pointer-analysis/value_set.cpp [hunk 2]
```diff
@@ -340,7 +340,7 @@ void value_sett::get_value_set(
 {
   exprt tmp(expr);
   if(!is_simplified)
-    simplify(tmp, ns);
+    simplifier(tmp, ns);
```
Comment (tautschnig) [inline src/pointer-analysis/value_set.cpp]: Why would one want a different simplifier? What would it look like? What are the minimal requirements?

File: src/pointer-analysis/value_set.cpp [hunk 3]
```diff
@@ -908,7 +908,7 @@ void value_sett::dereference_rec(
     dest=src;
 }
 
-void value_sett::get_reference_set(
+void value_sett::read_reference_set(
```
Comment (tautschnig) [inline src/pointer-analysis/value_set.cpp]: As above: why rename?

File: src/pointer-analysis/value_set.cpp [hunk 4]
```diff
@@ -1482,7 +1488,7 @@ void value_sett::do_end_function(
   assign(lhs, rhs, ns, false, false);
 }
 
-void value_sett::apply_code(
+void value_sett::apply_code_rec(
```
Comment (tautschnig) [inline src/pointer-analysis/value_set.cpp:1496]: As much as I agree this is a slightly better name: is the renaming essential?

File: src/pointer-analysis/value_set.cpp [hunk 5]
```diff
@@ -1609,6 +1615,10 @@ void value_sett::apply_code(
   {
     // doesn't do anything
   }
+  else if(statement==ID_dead)
+  {
+    // Ignore by default; could prune the value set.
+  }
```
Comment (tautschnig) [inline src/pointer-analysis/value_set.cpp:1626]: This is an unrelated change that would warrant a dedicated commit I'd say.

File: src/pointer-analysis/value_set.cpp [hunk 6]
```diff
@@ -1694,3 +1704,6 @@ exprt value_sett::make_member(
 
   return member_expr;
 }
+
+value_sett::expr_simplifiert value_sett::default_simplifier=
+  [](exprt &e, const namespacet &ns) { simplify(e, ns); };
```
Comment (tautschnig) [inline src/pointer-analysis/value_set.cpp]: See above: I have no idea what this is good for.

File: src/pointer-analysis/value_set.h [hunk 7]
```diff
@@ -24,8 +24,20 @@ class namespacet;
 
 class value_sett
 {
+  typedef std::function<void(exprt &, const namespacet &)> expr_simplifiert;
+
+  static expr_simplifiert default_simplifier;
```
Comment (tautschnig) [inline src/pointer-analysis/value_set.h]: See above: no clue.

File: src/pointer-analysis/value_set.h [hunk 8]
```diff
@@ -166,11 +178,20 @@ class value_sett
   typedef std::unordered_map<idt, entryt, string_hash> valuest;
   #endif
 
-  void get_value_set(
+  void read_value_set(
     const exprt &expr,
     value_setst::valuest &dest,
     const namespacet &ns) const;
 
+  /// Deprecated. Use the less-ambiguous read_value_set instead.
```
Comment (tautschnig) [inline src/pointer-analysis/value_set.h]: What's ambiguous about it? The signature appears to be the same?

File: src/pointer-analysis/value_set.h [hunk 9]
```diff
@@ -232,23 +256,25 @@ class value_sett
     const exprt &lhs,
     const namespacet &ns);
 
-  void get_reference_set(
+  void read_reference_set(
     const exprt &expr,
     value_setst::valuest &dest,
     const namespacet &ns) const;
 
+  /// Deprecated. Use the less-ambiguous read_reference_set instead.
```
Comment (tautschnig) [inline src/pointer-analysis/value_set.h]: As above.

File: src/pointer-analysis/value_set.h [hunk 10]
```diff
@@ -287,6 +306,67 @@ class value_sett
     const exprt &src,
     const irep_idt &component_name,
     const namespacet &ns);
+
+  // Expression simplification:
+
+private:
+  /// Expression simplification function; by default, plain old
+  /// util/simplify_expr, but can be customised by subclass.
+  expr_simplifiert simplifier;
```
Comment (tautschnig) [inline src/pointer-analysis/value_set.h]: A few answers to the "why" question would be appreciated.

File: src/pointer-analysis/value_set_domain_transform.inc [hunk 11]
```diff
@@ -25,13 +26,18 @@ void value_set_domaint::transform(
     break;
 
   case END_FUNCTION:
-    value_set.do_end_function(static_analysis_baset::get_return_lhs(to_l), ns);
+  {
+    value_set.do_end_function(
+      static_analysis_baset::get_return_lhs(to_l), ns);
```
Comment (tautschnig) [inline src/pointer-analysis/value_set_domain_transform.inc:31]: No actual changes here?

File: src/pointer-analysis/value_set_domain_transform.inc [hunk 12]
```diff
@@ -44,7 +50,8 @@ void value_set_domaint::transform(
       const code_function_callt &code=
         to_code_function_call(from_l->code);
 
-      value_set.do_function_call(to_l->function, code.arguments(), ns);
+      value_set.do_function_call(
+        to_l->function, code.arguments(), ns);
```
Comment (tautschnig) [inline src/pointer-analysis/value_set_domain_transform.inc:54]: Required?

**Review by tautschnig — COMMENTED** [review]
What I struggle with in this PR is the number of non-essential changes that make it much harder to understand the key ingredients. (Whether the introduction of templates is the right thing to do is a different question that I'm a bit unsure to answer for any use of `value_set_*` should rather be abstained from given its poor performance and the use of deprecated `static_analysist`.)

File: src/goto-symex/postcondition.cpp (hunk 1 above)
Comment (smowton) [reply src/goto-symex/postcondition.cpp]: These were renamed because in an earlier version of this PR, the 4-arg version of `get_value_set` was virtual but the 3-arg version wasn't, which was annoying as you had to manually forward it in each derived class, hence renaming the overload out of the way. However since that has changed, I'll remove these.

File: src/pointer-analysis/value_set.cpp (hunk 2 above)
Comment (smowton) [reply src/pointer-analysis/value_set.cpp]: It simplifies an `exprt &` in place, like `simplify`. We use it to turn off the default simplifier's merging of identical struct types.

File: src/pointer-analysis/value_set.cpp (hunk 4 above)
Comment (smowton) [reply src/pointer-analysis/value_set.cpp:1496]: The case for this was: it's better to rename this and override the private function than to override the public `apply_code`, thus separating the customisation point (`apply_code_rec`) from the public interface (`apply_code`).

File: src/pointer-analysis/value_set_domain_transform.inc (hunk 11 above)
Comment (smowton) [reply src/pointer-analysis/value_set_domain_transform.inc:31]: Yeah, the existing code spilled over the 80-column boundary so I got linter complaints on moving the file like this.

File: src/pointer-analysis/value_set_domain_transform.inc (hunk 12 above)
Comment (smowton) [reply src/pointer-analysis/value_set_domain_transform.inc:54]: As above.

**Comment by smowton** [comment]
@tautschnig removed the `read_*` renamings; replied to other comments with reasoning.

**Comment by smowton** [comment]
Also moved the "allow value_sett to handle DEAD if it wants to" change into a different commit.

**Review by peterschrammel — APPROVED** [review]

File: src/pointer-analysis/value_set.cpp (hunk 6 above)
Comment (peterschrammel) [reply src/pointer-analysis/value_set.cpp]: Also not sure. @smowton, do you have a use case for a custom simplifier?

Comment (smowton) [reply src/pointer-analysis/value_set.cpp]: Yes, to prevent merging of identical structs (e.g. two Java objects with the same field names but unrelated types having casts erased)

**Comment by smowton** [comment]
@tautschnig any further objections?

**Comment by smowton** [comment]
@tautschnig ping

File: src/pointer-analysis/value_set.cpp (hunk 6 above)
Comment (tautschnig) [reply src/pointer-analysis/value_set.cpp]: That's removing the symptom rather than the root cause, I'd say. If Java objects need to be handled differently (which part of the code base requires this?), then the Java front-end needs to make sure this happens. We can't go across all the code base and build in special cases in each place.

**Review by tautschnig — COMMENTED** [review]
My apologies for the long delay, but I continue to struggle with this custom-simplifier thing. I find this really broken, as outlined in my comment.

**Comment by smowton** [comment]
It's really quite hard to solve this in general. The problem is that Java stub classes (ones we don't have bytecode for) have an unknown inheritance graph, so you can get apparently unrelated types like
```
struct A {
  int x;
};
struct B {
  int x;
};
```
But in truth they are parent and child, and therefore legal to cast between. After `get_reference_set` or similar have been applied this results in struct casts, like `(struct A)dynamic_object1` (where `dynamic_object1` has type `B`). Finally `simplify_expr` removes typecasts where the operand is `base_type_eq` the result type, which removes casts between like-shaped structs, and you get a value-set with types that don't match the type of the expression you queried.

Either every query needs to copy with producing potentially wrongly-typed outputs, or the cast needs to be retained.

Possibilities:
(1) `simplify_expr` was wrong to ever change the type of the expression it's simplifying. It should only eliminate typecasts that are actually redundant, in that they cast an operand to the exact same type.
(1.5) Don't change type *when the struct type in question has type == ID_java* (or type != ID_C ... ?) This is hard to check since struct tags lack the `java::` prefix present in their symbol name. Add a new `ID_symbol_name` to `struct_typet` to ensure the ability to look them up in the symtab reliably?
(2) `value_sett` should re-add casts at the last stage of `get_value_set` so that querying `e` only ever returns expressions of matching type (with casts if necessary)
(3) Some users of `simplify_expr` want cast elimination like this and others don't. Add a flag like the existing `simplify_expr_classt::do_simplify_if` to customise what it is and isn't allowed to simplify. <-- the approach used here. I allowed customisation and only used the don't-eliminate-struct-casts option in the security product in case some part of goto-symex depends on this behaviour, or loses accuracy when it isn't present.
(4) `value_sett::get_value_set` should continue to return mixed-type value sets, and users should apply casts themselves.

Any preferences?

**Comment by tautschnig** [comment]
It seems that `simplify_expr` is possibly wrong here: it should not be removing type casts when those types aren't literally the same (i.e., comparison using `==` succeeds). That is, unless we forbid using `==` on `typet` and enforce using `base_type_eq` (which is semantic, but not syntactic equality). In the solver back end it may be safe to drop such casts, but this should not be done in any higher layer.

It should be a postcondition in `simplify_expr` that the final expression has literally the same type as the initial one.

@smowton I'm aware this is a fair bit of extra work, but one that is highly valuable for C code as well.

**Comment by smowton** [comment]
@tautschnig I tried just changing `base_type_eq` for `==`; this mostly worked, but broke some string solver tests, which I presume depend on a disappearing typecast. Therefore I've removed the simplifier stuff from here and will keep it out of tree in the security product until I can figure out how to fix the string solver. I haven't tried the broader postcondition yet but I will soon.

**Comment by smowton** [comment]
@tautschnig I've added another commit documenting `value_sett`. I can split this off into a separate later PR if you like, but since it's ready, concerns the same code and doesn't cause any functional change I thought it might be appropriate to include here.

**Comment by tautschnig** [comment]
It seems c67dc8a needs some more work to make the doxygen checks pass?

File: src/pointer-analysis/value_set.h [hunk 13]
```diff
@@ -298,30 +494,81 @@ class value_sett
...
+  virtual void apply_code_rec(
+    const codet &code,
+    const namespacet &ns);
+
+ private:
+  /// Subclass customisation point to filter or otherwise alter the value-set
+  /// returned from get_value_set before it is passed into assign. For example,
+  /// this is used in one subclass to tag and thus differentiate values that
+  /// originated in a particular place, vs. those that have been copied.
+  virtual void adjust_assign_rhs_values(
```
Comment (tautschnig) [inline src/pointer-analysis/value_set.h:554]: Why is this marked  `private` (yet `virtual`)?!

Comment (smowton) [reply src/pointer-analysis/value_set.h:554]: There's no base implementation for the subclass implementation to call, so no need to make it protected. Could make it so if you'd prefer subclasses to always call it in case it grows one, but any such base class behaviour could be directly implemented at the callsite.

**Comment by smowton** [comment]
@tautschnig Doxygen fixes applied
